snuba/pipeline/stages/query_execution.py

```diff
@@ -0,0 +1,358 @@
+        )
+        stats = update_query_metadata_and_stats(
+            query=clickhouse_query,
+            sql=formatted_query.get_sql(),
+            stats=stats,
+            query_metadata_list=query_metadata.query_list,
+            query_settings={},
+            trace_id="",
+            status=QueryStatus.INVALID_REQUEST,
+            request_status=get_request_status(cause),
+        )
+        raise QueryException.from_args(
+            cause.__class__.__name__,
+            str(cause),
+            extra=QueryExtraData(
+                stats=stats,
+                sql=formatted_sql,
+                experiments=clickhouse_query.get_experiments(),
+            ),
+        ) from cause
+    with sentry_sdk.start_span(description=formatted_sql, op="function") as span:
+        span.set_tag("table", table_names)
+
+        def execute() -> QueryResult:
+            return db_query(
+                clickhouse_query=clickhouse_query,
+                query_settings=query_settings,
+                attribution_info=attribution_info,
+                dataset_name=query_metadata.dataset,
+                formatted_query=formatted_query,
+                reader=reader,
+                timer=timer,
+                query_metadata_list=query_metadata.query_list,
+                stats=stats,
+                trace_id=span.trace_id,
+                robust=robust,
+            )
+
+        if concurrent_queries_gauge is not None:
+            with concurrent_queries_gauge:
+                return execute()
+        else:
+            return execute()
+
+
+def get_query_size_group(query_size_bytes: int) -> str:
+    """
+    Given the size of a query string in bytes, returns a string
+    representing the size of the query in 10% grouped increments
+    of the Maximum Query Size as defined in Snuba settings.
+
+    Eg. If the query size is 40-49% of the max query size, this function
+    returns ">=40%".
+
+    Eg. If the query size is equal to the max query size, this function
+    returns "100%".
+    """
+    if query_size_bytes == MAX_QUERY_SIZE_BYTES:
+        return "100%"
+    else:
+        query_size_group = int(floor(query_size_bytes / MAX_QUERY_SIZE_BYTES * 10)) * 10
+        return f">={query_size_group}%"
+
+
+def _apply_turbo_sampling_if_needed(
+    clickhouse_query: ClickhouseQuery | CompositeQuery[Table],
+    query_settings: QuerySettings,
+) -> None:
+    """
+    TODO: Remove this method entirely and move the sampling logic
+    into a query processor.
+    """
+    if isinstance(clickhouse_query, ClickhouseQuery):
+        if (
+            query_settings.get_turbo()
+            and not clickhouse_query.get_from_clause().sampling_rate
+        ):
+            clickhouse_query.set_from_clause(
+                replace(
+                    clickhouse_query.get_from_clause(),
+                    sampling_rate=snuba_settings.TURBO_SAMPLE_RATE,
+                )
+            )
```

Comment:
This is all copy-pasted from `snuba/web/query` to ensure that we are truly running two separate pipelines. open to just importing these functions but I like having the entire execution stage in one file. If this works, we can just remove all the duplicate code in `web/query.py`